src/lib/viewers/image/ImageViewer.js

```diff
@@ -165,6 +187,56 @@ class ImageViewer extends ImageBaseViewer {
         };
     }
 
+    /**
+     * Noop for handling annotation creation event.
+     *
+     * @override
+     * @return {void}
+     */
+    handleAnnotationCreateEvent() {}
+
+    /**
+     * Noop for handling annotation creator change event.
+     *
+     * @override
+     * @return {void}
+     */
+    handleAnnotationCreatorChangeEvent() {}
+
+    /**
+     * Gets the viewport dimensions.
+     *
+     * @return {Object} the width & height of the viewport
+     */
+    getViewportDimensions() {
+        return {
+            width: this.wrapperEl.clientWidth - 2 * IMAGE_PADDING,
+            height: this.wrapperEl.clientHeight - 2 * IMAGE_PADDING,
+        };
+    }
+
+    /**
+     * Sets mode to be AnnotationMode.NONE if the zoomed image overflows the viewport.
+     *
+     * @return {void}
+     */
+    handleZoomEvent({ newScale, type }) {
+        const [width, height] = newScale;
+        const isUserInitiated = type !== undefined;
+        if (!isUserInitiated) {
+            return;
+        }
+
+        const viewport = this.getViewportDimensions();
+
+        // We only set AnnotationMode to be NONE if the image overflows the viewport and the state is not explicit region creation
+        const currentState = this.annotationControlsFSM.getState();
+        if (currentState === AnnotationState.REGION_TEMP && (width > viewport.width || height > viewport.height)) {
```

Comment:
So if user explicitly turns off region mode, the state in FSM will become `NONE` and the mode in annotations will become `REGION`, and then user zoom in, but this will not work anymore, because the `currentState` is `NONE`. Then user cannot "pan" anymore.

Reply:
Yes I see what you mean. Maybe I can do this:

```
if ((currentState === AnnotationState.REGION_TEMP || currentState === AnnotationState.NONE) && (width > viewport.width || height > viewport.height)) {
    this.annotator.toggleAnnotationMode(AnnotationMode.NONE);
    this.processAnnotationModeChange(this.annotationControlsFSM.transition(AnnotationInput.CANCEL));
}
```

Reply:
Good callout @mxiao6, we probably need to account for the button click -- probably need to override `handleAnnotationControlsClick` for how the `ImageViewer` experience should be

Reply:
Since @ConradJChan merged his refactor change, I think we no longer need to override `handleAnnotationControlsClick`.